applications/OptimizationApplication/custom_utilities/filtering/explicit_vertex_morphing_filter.cpp

```diff
@@ -0,0 +1,342 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   license: OptimizationApplication/license.txt
+//
+//  Main authors:    Baumgaertner Daniel
+//                   Aditya Ghantasala
+//                   Suneth Warnakulasuriya
+//
+
+// System includes
+#include <sstream>
+#include <type_traits>
+#include <tuple>
+
+// Project includes
+#include "expression/literal_flat_expression.h"
+#include "utilities/builtin_timer.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/atomic_utilities.h"
+
+// Application includes
+
+// Include base h
+#include "explicit_vertex_morphing_filter.h"
+
+namespace Kratos {
+
+namespace ExplicitVertexMorphingFilterHelperUtilities
+{
+
+template<class TContainerType>
+const TContainerType& GetContainer(const ModelPart& rModelPart);
+
+template<>
+const ModelPart::NodesContainerType& GetContainer(const ModelPart& rModelPart)
+{
+    return rModelPart.Nodes();
+}
+
+template<>
+const ModelPart::ConditionsContainerType& GetContainer(const ModelPart& rModelPart)
+{
+    return rModelPart.Conditions();
+}
+
+template<>
+const ModelPart::ElementsContainerType& GetContainer(const ModelPart& rModelPart)
+{
+    return rModelPart.Elements();
+}
+
+template<class TContainerType>
+Expression::ConstPointer GetNodalDomainSizeExpression(
+    const TContainerType& rContainer,
+    const ModelPart::NodesContainerType& rNodes)
+{
+    const IndexType number_of_nodes = rNodes.size();
+
+    auto p_expression = LiteralFlatExpression<double>::Create(number_of_nodes, {});
+
+    // reset the existing values in the expression
+    IndexPartition<IndexType>(number_of_nodes).for_each([&p_expression](const IndexType Index) {
+        *(p_expression->begin() + Index) = 0.0;
+    });
+
+    IndexPartition<IndexType>(rContainer.size()).for_each([&rNodes, &rContainer, &p_expression](const IndexType Index) {
+        const auto& r_geometry = (rContainer.begin() + Index)->GetGeometry();
+        const double nodal_domain_size = r_geometry.DomainSize() / r_geometry.size();
+        for (const auto& r_node : r_geometry) {
+            const auto node_ptr = rNodes.find(r_node.Id());
+            if (node_ptr != rNodes.end()) {
+                const IndexType node_index = std::distance(rNodes.begin(), node_ptr);
+                AtomicAdd<double>(*(p_expression->begin() + node_index), nodal_domain_size);
+            } else {
+                KRATOS_ERROR << "Node with id " << r_node.Id() << " not found.";
+            }
+        }
+    });
+
+    return p_expression;
+}
+
+template<class TEntityType>
+double GetDomainSize(
+    const EntityPoint<TEntityType>& rPoint,
+    Expression const * const pExpression)
+{
+    return rPoint.GetEntity().GetGeometry().DomainSize();
```

Comment:
I am concerned here. If we have a nodalexpression to filter, then we do node search on the nodes to find the neighbors and weights. Then we need to apply integration, depending the field is integrated or not,  we need to get the NODAL area or volume as the integration weight. However, DomainSize does the job in case the expression is an element entity but what about elemental/conditional expressions?

Reply:
In the condition and element case, the domain size is used. In the specialized node case, the representative nodal area/volumes are used.